browser: implement  `page.goBack()` and  `page.goForward()`

pkalsi97 opened the pull request:
## What?
This PR implements `page.goBack()` and `page.goForward()` to allow users to navigate through the browser's session history. 

I have kept the implementation inline with `Reload()` and followed guidelines mentioned in the issue.
## Why?
This functionality was missing and users had to rely on `page.evaluate(() => window.history.back())`, which is racy and unreliable for testing purposes as mentioned in the issue #5400


#### Example
```javascript
import { browser } from 'k6/browser';
import { check } from 'k6';

export const options = {
    scenarios: {
        ui: {
            executor: 'shared-iterations',
            options: {
                browser: {
                    type: 'chromium',
                },
            },
        },
    },
};

export default async function () {
    const page = await browser.newPage();

    try {
        await page.goto('data:text/html,<h1>Page 1</h1>');
        await page.goto('data:text/html,<h1>Page 2</h1>');

        await page.goBack();
        const url1 = await page.url();
        check(url1, {
            'goBack navigated to Page 1': (u) => u.includes('Page 1') || u.includes('Page%201'),
        });

        await page.goForward();
        const url2 = await page.url();
        check(url2, {
            'goForward navigated to Page 2': (u) => u.includes('Page 2') || u.includes('Page%202'),
        });

        const nullResp = await page.goForward();
        check(nullResp, {
            'goForward at boundary returns null': (r) => r === null,
        });

    } finally {
        await page.close();
    }
}
``` 
## Checklist
- [x] I have performed a self-review of my code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests for my changes.
- [x] I have run linter and tests locally (`make check`) and all pass.

## Checklist: Documentation (only for k6 maintainers and if relevant)
**Please do not merge this PR until the following items are filled out.**
- [ ] I have added the correct milestone and labels to the PR.
- [ ] I have updated the release notes: _link_
- [ ] I have updated or added an issue to the [k6-documentation](https://github.com/grafana/k6-docs): grafana/k6-docs#NUMBER if applicable
- [ ] I have updated or added an issue to the [TypeScript definitions](https://github.com/grafana/k6-DefinitelyTyped/tree/master/types/k6): grafana/k6-DefinitelyTyped#NUMBER if applicable

<!-- - [ ] Any other relevant item -->

## Related PR(s)/Issue(s)
Closes #5400

inancgumus commented on internal/js/modules/k6/browser/common/page.go:
```diff
@@ -1490,6 +1490,220 @@ func (p *Page) Reload(opts *PageReloadOptions) (_ *Response, rerr error) { //nol
+		return nil, histErr
+	}
+
+	// check if we can goBack
```
Please remove, as the code is clear.

inancgumus commented on internal/js/modules/k6/browser/common/page.go:
```diff
@@ -1490,6 +1490,220 @@ func (p *Page) Reload(opts *PageReloadOptions) (_ *Response, rerr error) { //nol
+
+	navAction := page.NavigateToHistoryEntry(targetEntryID)
+	if err := navAction.Do(cdp.WithExecutor(p.ctx, p.session)); err != nil {
+		return nil, err
```
Please add an error explanation as I suggested earlier.

inancgumus commented on internal/js/modules/k6/browser/tests/page_test.go:
```diff
@@ -3895,3 +3895,95 @@ func TestPageUnrouteAll(t *testing.T) {
+		tb := newTestBrowser(t, withFileServer())
+		p := tb.NewPage(nil)
+
+		// Navigate to first page
```
Please remove.

pkalsi97 commented:
@inancgumus, I have implemented all the suggestions, thanks for pointing out a critical issue with the implementation.
1. Fixed the `GoBack` and `GoForward`
2. Removed unnecessary comments
3. Refactored integration test to use helper functions
4. Added iframe integration test for `GoBack` and `GoForward`

#### What ?
`GoBack` and `GoForward` methods were timing out when navigating between real HTML pages, similar was observed with iframes. 

#### FIX
Changed the implementation to **poll for URL change** instead of waiting for `EventFrameNavigation`:
1. Get the target URL from the browser's navigation history (we know where we're going)
2. Execute `NavigateToHistoryEntry()`  
3. Poll every 50ms checking if `mainFrame.URL() == targetURL`

Also I am not sure why 2 test were failing in ci. please guide me here.

inancgumus commented on internal/js/modules/k6/browser/common/page_options.go:
```diff
@@ -33,6 +33,12 @@ type PageScreenshotOptions struct {
 
+// PageGoBackForwardOptions are options for Page.GoBack and Page.GoForward.
+type PageGoBackForwardOptions struct {
+	WaitUntil LifecycleEvent `json:"waitUntil" js:"waitUntil"`
```
Do we need `js:"..."` here? What's the reason `WaitUntil` has it but not `Timeout`?

inancgumus commented on internal/js/modules/k6/browser/common/page_options.go:
```diff
@@ -67,6 +73,39 @@ func NewPageReloadOptions(defaultWaitUntil LifecycleEvent, defaultTimeout time.D
+// Parse parses the page go back/forward options.
+func (o *PageGoBackForwardOptions) Parse(ctx context.Context, opts sobek.Value) error {
+	rt := k6ext.Runtime(ctx)
+	if !common.IsNullish(opts) {
```
Please use early return here:

```go
if common.IsNullish(opts) {
    return nil
}
...
```

We also don't need `rt` to be a variable and we can pass it directly to `ToObject`.

pkalsi97 commented:
@inancgumus  Done.

inancgumus commented:
@pkalsi97, it would be great if you could provide updates to k6-docs and TypeScript definitions for this feature. Thanks!

pkalsi97 commented on internal/js/modules/k6/browser/common/page.go:
```diff
@@ -1490,6 +1490,87 @@ func (p *Page) Reload(opts *PageReloadOptions) (_ *Response, rerr error) { //nol
+	}
+
+	// Poll for URL change, don't rely on lifecycle events since bfcache
+	// restorations don't re-fire them.
```
I'll also look into it and figure it out.

codebien commented:
@pkalsi97 @inancgumus I moved this pull request to v1.7.0 release. If in the meantime we manage to fix the conflicts then we'll move back to the upcoming v1.6.0 release.
